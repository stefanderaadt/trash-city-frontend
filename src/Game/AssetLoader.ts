import * as PIXI from 'pixi.js';
import axios from 'axios';
import { parseStringPromise } from 'xml2js';

import store from '../store';

import { startLoading, onLoading, loadingCompleted } from '../actions/loaderActions';

import spritesheetTilesetMap from './assets/spritesheet_tileset_map.json';
import soundMap from './assets/sound_map.json';

class AssetLoader {
  private loader: PIXI.Loader;
  private tilesets: object = {};

  public init = async (): Promise<void> => {
    // Get new PIXI Loader
    this.loader = new PIXI.Loader();

    // Set events
    this.loader.onStart.add(this.onStart);
    this.loader.onLoad.add(this.onLoad);
    this.loader.onComplete.add(this.onComplete);

    // Map 'TexturePacker' ids to 'Tiled' tile ids and add 'TexturePacker' tilesets to loader.
    // These tile maps are based on '/src/Game/assets/spritesheet_tileset_map.json'.
    // To tell Pixi what sprites to load and to retrieve the right sprites from created maps in 'Tiled'.
    await spritesheetTilesetMap.forEach(async map => {
      // Add 'TexturePacker' tileset to loader
      this.loader.add(map.spritesheet);
      const tileset = await axios.get(map.tileset);

      // Parse 'Tiled' tileset XML data
      const xml = await parseStringPromise(tileset.data);

      // Map 'TexturePacker' ids to 'Tiled' tile ids
      this.tilesets[map.spritesheet] = xml.tileset.tile.map(t => ({
        mapId: parseInt(t.$.id, 10),
        id: t.image[0].$.source
      }));
    });

    soundMap.forEach(map => {
      this.loader.add(map.key, map.path);
    });

    // Start loading process
    this.loader.load();
  };

  private onStart = (): void => {
    // Dispatch loading process to front-end
    store.dispatch(startLoading());
  };

  private onLoad = (e): void => {
    // Dispatch update front-end with new progress (for loading bar)
    store.dispatch(onLoading(e.progress));
  };

  private onComplete = (): void => {
    // Dispatch loadingCompleted event to front-end
    store.dispatch(loadingCompleted());
  };

  // Get assets by 'Tiled' tile ids for tilemap
  public getAssetByMapId = ({ tileset, mapId }): PIXI.Texture => {
    const tileSetMap = this.tilesets[tileset];
    const tile = tileSetMap.find(t => t.mapId === mapId);

    if (!tile) throw new Error(`Tile with mapId: ${mapId} not found!`);

    return this.getAssetById({ tileset, id: tile.id });
  };

  // Get loaded sound by ID to play in gameobject
  public getSoundById = ({ id }): PIXI.sound.Sound =>
    this.loader?.resources[id].sound

  // Get assets by 'Texture Packer' id
  public getAssetById = ({ tileset, id }): PIXI.Texture =>
    this.loader?.resources[tileset]?.spritesheet?.textures[id];

  // Get animations by 'Texture Packer' id
  public getAnimationById = ({ tileset, id }): object =>
    this.loader?.resources[tileset]?.spritesheet?.animations[id];
}

export default AssetLoader;
