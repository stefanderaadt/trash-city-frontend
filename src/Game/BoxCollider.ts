import Position from './gameObjects/Position';

class BoxCollider {
  public position: Position;
  public width: number;
  public height: number;
  public halfWidth: number;
  public halfHeight: number;

  constructor({ width, height, position }) {
    this.width = width;
    this.height = height;
    this.position = position;

    this.halfWidth = width / 2;
    this.halfHeight = height / 2;
  }

  // Check if other BoxCollider is in radius of this BoxCollider with variable radius
  public isInRadius = (otherBoxCollider: BoxCollider, radius: number): boolean => {
    return (
      (otherBoxCollider.position.x - this.position.x) *
        (otherBoxCollider.position.x - this.position.x) +
        (otherBoxCollider.position.y - this.position.y) *
          (otherBoxCollider.position.y - this.position.y) <=
      radius * radius
    );
  };
}

export default BoxCollider;
