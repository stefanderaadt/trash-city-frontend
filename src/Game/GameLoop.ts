class GameLoop {
  private isPlaying: boolean = false;
  private tref;
  private callback: Function;
  public frame: number = -1;
  public fps: number;

  constructor(fps: number, callback: Function) {
    this.fps = fps;
    this.callback = callback;
  }

  // Start gameloop with requestAnimationFrame
  public start = (): void => {
    if (!this.isPlaying) {
      this.isPlaying = true;
      this.tref = requestAnimationFrame(this.loop);
    }
  };

  // Pause gameloop
  public pause = (): void => {
    if (this.isPlaying) {
      cancelAnimationFrame(this.tref);
      this.isPlaying = false;
    }
  };

  // Reset gameloop
  public reset = (): void => {
    this.pause();
    this.frame = -1;
  };

  // Actual gameloop with callback
  private loop = (): void => {
    this.frame++;
    this.tref = requestAnimationFrame(this.loop);
    this.callback(this.frame);
  };
}

export default GameLoop;
