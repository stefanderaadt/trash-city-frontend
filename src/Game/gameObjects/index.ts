/* eslint-disable import/no-cycle */
import Player from './objects/Player';
import Civilian from './objects/Civilian';
import CivilianSpawnPoint from './objects/CivilianSpawnPoint';
import Trash from './objects/Trash';
import WaterTile from './objects/WaterTile';
import FilterSystem from './objects/FilterSystem';

// Export all object constructors for GameObjectFactory
export default {
  Player,
  Civilian,
  CivilianSpawnPoint,
  Trash,
  WaterTile,
  FilterSystem
};
