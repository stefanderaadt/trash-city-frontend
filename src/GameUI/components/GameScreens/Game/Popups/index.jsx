import React from 'react';
import { connect } from 'react-redux';

import { setScreen } from '../../../../../actions/appActions';
import { setScoreName, saveHighScore } from '../../../../../actions/scoreActions';
import { Popups } from '../../../../../enums.ts';

import Pause from './Pause';
import Score from './Score';
import Info from './Info';

@connect(
  state => ({
    game: state.game,
    level: state.level,
    score: state.score
  }),
  dispatch => ({
    dispatch
  })
)
class GamePopups extends React.Component {
  setScreen = screen => {
    const { dispatch } = this.props;
    dispatch(setScreen(screen));
  };

  setScoreName = name => {
    const { dispatch } = this.props;
    dispatch(setScoreName(name));
  };

  saveHighScore = () => {
    const { dispatch } = this.props;
    dispatch(saveHighScore());
  };

  render() {
    const { game, level, score } = this.props;

    switch (game.currentPopup) {
      case Popups.Pause:
        return <Pause setScreen={this.setScreen} />;
      case Popups.Score:
        return (
          <Score
            setScreen={this.setScreen}
            game={game}
            score={score}
            setScoreName={this.setScoreName}
            saveHighScore={this.saveHighScore}
          />
        );
      case Popups.Info:
        return <Info level={level} />;
      default:
        return null;
    }
  }
}

export default GamePopups;
