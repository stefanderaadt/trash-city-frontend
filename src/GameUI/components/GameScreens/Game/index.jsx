import React from 'react';
import { connect } from 'react-redux';

import Button from '../../Default/Button';
import ProgressBar from '../../Default/ProgressBar';
import game from '../../../../Game/Game.ts';
import GamePopups from './Popups';
import displayTime from '../../../../utils/displayTime';
import numberFormatter from '../../../../utils/numberFormatter';

@connect(
  state => ({
    game: state.game
  }),
  dispatch => ({
    dispatch
  })
)
class Game extends React.Component {
  render() {
    const {
      game: { score, time, coins }
    } = this.props;

    let icon = '';

    if (score < 20) {
      icon = 'icon-astonished-face';
    } else if (score < 40) {
      icon = 'icon-sad-face';
    } else if (score < 60) {
      icon = 'icon-neutral-face';
    } else if (score < 80) {
      icon = 'icon-smiling-face';
    } else {
      icon = 'icon-laughing-face';
    }

    return (
      <div className="game">
        <div className="game-overlay-ui">
          <div className="top-bar">
            <Button onClick={game.pause}>
              <i className="icon-pause" />
            </Button>
          </div>
          <div className="bottom-bar">
            <div className="bottom-bar-progress">
              <div className="bottom-bar-progress-bar">
                <ProgressBar progress={score} description={v => `${numberFormatter(v)}%`} />
              </div>
              <div className="bottom-bar-progress-icon">
                <i className={icon} />
              </div>
            </div>
            <div className="bottom-bar-info">
              {`Muntjes: ${coins} - Tijd: ${displayTime(time)}`}
            </div>
          </div>
        </div>
        <GamePopups />
      </div>
    );
  }
}

export default Game;
