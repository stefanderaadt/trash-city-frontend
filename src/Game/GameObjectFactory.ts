/* eslint-disable import/no-cycle */
import gameObjects from './gameObjects';

import IGameObject from './gameObjects/IGameObject';

class GameObjectFactory {
  // Create new GameObject instance based on name
  public create = ({ name, ...props }): IGameObject => {
    // Check if GameObject with name exists
    if (!(name in gameObjects)) throw new Error(`gameObject with name: ${name} not found.`);

    // Create instance of GameObject and pass properties
    return new gameObjects[name](props);
  };
}

export default GameObjectFactory;
