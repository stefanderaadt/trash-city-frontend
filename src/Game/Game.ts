/* eslint-disable import/no-cycle */
import * as PIXI from 'pixi.js';

import store from '../store';
import { setScreen } from '../actions/appActions';
import { Screens, Popups } from '../enums';
import GameLoop from './GameLoop';
import levels from './levels/levels';
import IConfigLevel from './levels/IConfigLevel';
import AssetLoader from './AssetLoader';
import GameObjectFactory from './GameObjectFactory';
import Level from './Level';
import InputHandler from './InputHandler';
import SoundManager from './SoundManager';

import {
  pauseGame,
  resumeGame,
  setTime,
  setScore,
  setCoins,
  showPopup,
  hidePopup
} from '../actions/gameActions';
import { levelInit } from '../actions/levelActions';

const size = [960, 512];
const fps = 60;

class Game {
  public gameLoop: GameLoop;
  public game: PIXI.Application;
  public assetLoader: AssetLoader;
  public gameObjectFactory: GameObjectFactory;
  public currentLevel: Level;
  public inputHandler: InputHandler;
  public soundManager: SoundManager;
  public coins: number = 0;

  public async init(): Promise<void> {
    const gameId = document.getElementById('game');

    this.game = new PIXI.Application({
      backgroundColor: 0x1099bb
    });

    // PIXI Settings
    PIXI.settings.SCALE_MODE = PIXI.SCALE_MODES.NEAREST; // Display pixelart correctly

    // Auto resize window
    this.resize();
    window.addEventListener('resize', this.resize);

    // Append game canvas to html element with '#game'
    if (gameId) gameId.appendChild(this.game.view);

    this.assetLoader = new AssetLoader();
    await this.assetLoader.init();

    this.gameObjectFactory = new GameObjectFactory();

    this.inputHandler = new InputHandler();

    this.soundManager = new SoundManager();

    this.gameLoop = new GameLoop(fps, this.loop);
  }

  // Picked level from front-end
  public pickLevel = (id: number): void => {
    this.gameLoop.reset();

    // Set front-end screen to game
    store.dispatch(setScreen(Screens.Game));
    // Tell front-end that the game has started
    store.dispatch(resumeGame());

    // Get level config by given id
    const foundLevel: IConfigLevel = levels.find(l => l.id === id) as IConfigLevel;

    // Create new level
    this.currentLevel = new Level();
    // Initialize level and pass level config
    this.currentLevel.init(foundLevel);

    // Clear coins
    this.setCoins(0);

    // Save level data to front-end and display info popup
    store.dispatch(
      levelInit({ name: foundLevel.name, goal: foundLevel.goal, levelId: foundLevel.id })
    );
  };

  // Hide info popup and start game
  public startGame = (): void => {
    store.dispatch(hidePopup());
    this.gameLoop.start();
  };

  // Pause all layers/objects and display pause popup
  public pause = (): void => {
    this.gameLoop.pause();
    this.currentLevel.pause();
    store.dispatch(pauseGame());
  };

  // Resume all layers/objects and hide pause popup
  public resume = (): void => {
    this.gameLoop.start();
    this.currentLevel.resume();
    store.dispatch(resumeGame());
  };

  // Update score to front-end
  public setScore = (score: number): void => {
    store.dispatch(setScore(score));

    // Show score popup when the user reached 100%
    if (score >= 100) {
      this.pause();
      store.dispatch(showPopup(Popups.Score));
    }
  };

  // Get current game score
  public getScore = (): number => {
    return store.getState().game.score;
  };

  // Add coins to the current players coins
  public addCoins = (coins): void => {
    this.setCoins(this.coins + coins);
  };

  // Remove coins from the current player coins
  public removeCoins = (coins): void => {
    this.setCoins(this.coins - coins);
  };

  // Update variable and send action to UI
  public setCoins = (coins): void => {
    this.coins = coins;
    store.dispatch(setCoins(coins));
  };

  private loop = (frame: number): void => {
    // Loop callback from GameLoop
    this.update(frame);
    this.render();
  };

  private update = (frame): void => {
    // Update level
    this.currentLevel.update();

    // Update time every second
    if (frame % fps === 0) {
      store.dispatch(setTime(frame / fps));
    }
  };

  private render = (): void => {
    // Render level
    this.currentLevel.render();
  };

  // Update canvas content position after resize
  private resize = (): void => {
    const vpw: number = window.innerWidth;
    const vph: number = window.innerHeight;
    let nvw: number;
    let nvh: number;

    if (vph / vpw < size[1] / size[0]) {
      nvh = vph;
      nvw = (nvh * size[0]) / size[1];
    } else {
      nvw = vpw;
      nvh = (nvw * size[1]) / size[0];
    }

    this.game.renderer.resize(nvw, nvh);

    this.game.stage.scale.set(nvw / size[0], nvh / size[1]);
  };
}

// Only create one game instance
export default new Game();
