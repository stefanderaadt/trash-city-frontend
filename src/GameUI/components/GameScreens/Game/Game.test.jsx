/* eslint-disable no-console */
import React from 'react';
import { render } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import Game from './index';

const mockStore = configureStore([]);

jest.mock('../../../../Game/Game.ts', () => {
  return {
    A: true,
    B: false
  };
});

describe('Game tests', () => {
  test('Game rendering correctly with lowest score', () => {
    const store = mockStore({
      game: {
        score: 0,
        time: 300
      }
    });

    const wrapper = render(
      <Provider store={store}>
        <Game />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('Game rendering correctly with low score', () => {
    const store = mockStore({
      game: {
        score: 30,
        time: 300
      }
    });

    const wrapper = render(
      <Provider store={store}>
        <Game />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('Game rendering correctly with normal score', () => {
    const store = mockStore({
      game: {
        score: 50,
        time: 300
      }
    });

    const wrapper = render(
      <Provider store={store}>
        <Game />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('Game rendering correctly with high score', () => {
    const store = mockStore({
      game: {
        score: 70,
        time: 300
      }
    });

    const wrapper = render(
      <Provider store={store}>
        <Game />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('Game rendering correctly with highest score', () => {
    const store = mockStore({
      game: {
        score: 100,
        time: 300
      }
    });

    const wrapper = render(
      <Provider store={store}>
        <Game />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
