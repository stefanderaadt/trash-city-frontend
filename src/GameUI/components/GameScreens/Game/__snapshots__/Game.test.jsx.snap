// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game tests Game rendering correctly with high score 1`] = `
<div
  class="game"
>
  <div
    class="game-overlay-ui"
  >
    <div
      class="top-bar"
    >
      <div
        class="button undefined"
        role="button"
        tabindex="0"
      >
        <i
          class="icon-pause"
        />
      </div>
    </div>
    <div
      class="bottom-bar"
    >
      <div
        class="bottom-bar-progress"
      >
        <div
          class="bottom-bar-progress-bar"
        >
          <div
            class="progress-bar"
          >
            <div
              class="progress-bar-display"
              style="width:70%"
            />
            <div
              class="progress-bar-text"
            >
              70%
            </div>
          </div>
        </div>
        <div
          class="bottom-bar-progress-icon"
        >
          <i
            class="icon-smiling-face"
          />
        </div>
      </div>
      <div
        class="bottom-bar-time"
      >
        Tijd: 05:00
      </div>
    </div>
  </div>
</div>
`;

exports[`Game tests Game rendering correctly with highest score 1`] = `
<div
  class="game"
>
  <div
    class="game-overlay-ui"
  >
    <div
      class="top-bar"
    >
      <div
        class="button undefined"
        role="button"
        tabindex="0"
      >
        <i
          class="icon-pause"
        />
      </div>
    </div>
    <div
      class="bottom-bar"
    >
      <div
        class="bottom-bar-progress"
      >
        <div
          class="bottom-bar-progress-bar"
        >
          <div
            class="progress-bar"
          >
            <div
              class="progress-bar-display"
              style="width:100%"
            />
            <div
              class="progress-bar-text"
            >
              100%
            </div>
          </div>
        </div>
        <div
          class="bottom-bar-progress-icon"
        >
          <i
            class="icon-laughing-face"
          />
        </div>
      </div>
      <div
        class="bottom-bar-time"
      >
        Tijd: 05:00
      </div>
    </div>
  </div>
</div>
`;

exports[`Game tests Game rendering correctly with low score 1`] = `
<div
  class="game"
>
  <div
    class="game-overlay-ui"
  >
    <div
      class="top-bar"
    >
      <div
        class="button undefined"
        role="button"
        tabindex="0"
      >
        <i
          class="icon-pause"
        />
      </div>
    </div>
    <div
      class="bottom-bar"
    >
      <div
        class="bottom-bar-progress"
      >
        <div
          class="bottom-bar-progress-bar"
        >
          <div
            class="progress-bar"
          >
            <div
              class="progress-bar-display"
              style="width:30%"
            />
            <div
              class="progress-bar-text"
            >
              30%
            </div>
          </div>
        </div>
        <div
          class="bottom-bar-progress-icon"
        >
          <i
            class="icon-sad-face"
          />
        </div>
      </div>
      <div
        class="bottom-bar-time"
      >
        Tijd: 05:00
      </div>
    </div>
  </div>
</div>
`;

exports[`Game tests Game rendering correctly with lowest score 1`] = `
<div
  class="game"
>
  <div
    class="game-overlay-ui"
  >
    <div
      class="top-bar"
    >
      <div
        class="button undefined"
        role="button"
        tabindex="0"
      >
        <i
          class="icon-pause"
        />
      </div>
    </div>
    <div
      class="bottom-bar"
    >
      <div
        class="bottom-bar-progress"
      >
        <div
          class="bottom-bar-progress-bar"
        >
          <div
            class="progress-bar"
          >
            <div
              class="progress-bar-display"
              style="width:0%"
            />
            <div
              class="progress-bar-text"
            >
              0%
            </div>
          </div>
        </div>
        <div
          class="bottom-bar-progress-icon"
        >
          <i
            class="icon-astonished-face"
          />
        </div>
      </div>
      <div
        class="bottom-bar-time"
      >
        Tijd: 05:00
      </div>
    </div>
  </div>
</div>
`;

exports[`Game tests Game rendering correctly with normal score 1`] = `
<div
  class="game"
>
  <div
    class="game-overlay-ui"
  >
    <div
      class="top-bar"
    >
      <div
        class="button undefined"
        role="button"
        tabindex="0"
      >
        <i
          class="icon-pause"
        />
      </div>
    </div>
    <div
      class="bottom-bar"
    >
      <div
        class="bottom-bar-progress"
      >
        <div
          class="bottom-bar-progress-bar"
        >
          <div
            class="progress-bar"
          >
            <div
              class="progress-bar-display"
              style="width:50%"
            />
            <div
              class="progress-bar-text"
            >
              50%
            </div>
          </div>
        </div>
        <div
          class="bottom-bar-progress-icon"
        >
          <i
            class="icon-neutral-face"
          />
        </div>
      </div>
      <div
        class="bottom-bar-time"
      >
        Tijd: 05:00
      </div>
    </div>
  </div>
</div>
`;
